### PR TITLE
refactor: replace command inputs and options by a dictionary-like high-level data abstraction

### PR DESCRIPTION
**Pull request by micalevisk**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

I notice that all operations on command's inputs and options are using `.find()` to search for the first input/option found

## What is the new behavior?

replace the usage of `Input[]` from `.command.ts` and `.action.ts` files by a new abstraction called `CommandContext` which hides the complexity of registering and retrieving some input/operation in a command lifecycle.

this change is intended to improve:

- code readability  
Instead of having a bunch of calls like `inputs.find((o) => o.name === 'foo')` we now have just one `inputs.resolveInput('foo')`, which is also slightly fast as it uses JS `Map` under the hood. This make it even clear that we shouldn't be define the same input name more than one time. I didn't write a validation for this scenario for [micro]performance sake but we could easily introduce that if needed
- type safety  
Instead of relying on type assertion and non-null assertions, we're using generics to make the code less error prone at compile time. The present solution doesn't differ that much from type assertions but there is a room to improve that easily by now

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
I'd love to finally get rid of these confusing Input/Output abstractions - these are leftovers from the first version of the CLI 😄

**Comment by micalevisk** [comment]
in order to try this version better, I've published this branch's version as  
`micalevisk-nestjs-cli` package.

File: commands/command-storage.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
+export interface CommandStorageEntry<
+  TValue extends boolean | string = boolean | string,
+> {
+  readonly name: string;
+  readonly value: TValue;
+  readonly options?: any;
+}
+
+export class CommandStorage {
```
Comment (kamilmysliwiec) [inline commands/command-storage.ts:9]: very nitpicky but command storage name makes me think it somewhat holds the "command" reference while this isn't true (it represents the command context/options)

Comment (micalevisk) [reply commands/command-storage.ts:9]: good point

naming is hard, you know haha

I'll think of another name for it

**Comment by kamilmysliwiec** [comment]
This introduces a regression when you run `nest new` without passing the application name as an argument

**Comment by micalevisk** [comment]
I'll see if I can introduce this change in the current version of the main instead

**Pull request closed without merging**